cpp/src/join/detail/traversal.cuh

```diff
@@ -0,0 +1,120 @@
+/*
```

Comment:
This header contains an implementation, and it appears to only be used in one place in this PR. Will be it included from other files in future PRs? If not, could the implementation just move into the .cu file that includes it?

Reply:
My two motivations for `intersection.cuh` and `traversal.cuh`:
1. Could be reused in other poly/quad traversal algorithms?
2. Avoid a 700+ line `quadtree_poly_filtering.cu` 

I'm not sure whether these can be reused, or if they're too specific, but my instinct was to factor them out by default.